marimbabot_vision/src/vision_node.py

```diff
@@ -1,25 +1,85 @@
 #!/usr/bin/env python
+
+import numpy as np
 import rospy
+import torch
+from PIL import Image as PILImage
+from sensor_msgs.msg import Image as ROSImage
 from std_msgs.msg import String
-from sensor_msgs import Image
+from transformers import (DonutProcessor, VisionEncoderDecoderConfig,
+                          VisionEncoderDecoderModel)
 
-def callbackImage(data):    # TODO
-    pass
-#  rospy.loginfo()
-    
-def listener():
+def detect_notes(img_array: np.array, pre_processor: DonutProcessor, model: VisionEncoderDecoderModel):
+    # Load image
+    image = PILImage.open(img_array).convert('RGB')
+
+    # Rotate image if needed
+    if image.size[0] > image.size[1]:
+        image = image.transpose(PILImage.Transpose.ROTATE_90)
+
+    # Generate initial sequence
+    decoder_input_ids = pre_processor.tokenizer(
+        "<s>",
+        add_special_tokens=False,
+        return_tensors="pt").input_ids
+
+    # Preprocess image
+    pixel_values = pre_processor(image, return_tensors="pt").pixel_values
+
+    # Run the model
+    outputs = model.generate(
+        pixel_values.to(device),
+        decoder_input_ids=decoder_input_ids.to(device),
+        max_length=model.decoder.config.max_position_embeddings,
+        early_stopping=True,
+        pad_token_id=pre_processor.tokenizer.pad_token_id,
+        eos_token_id=pre_processor.tokenizer.eos_token_id,
+        use_cache=True,
+        num_beams=1,
+        bad_words_ids=[[pre_processor.tokenizer.unk_token_id]],
+        return_dict_in_generate=True,
+    )
+
+    # Decode tokens
+    sequence = pre_processor.batch_decode(outputs.sequences)[0][4:]
+
+    # Clean output sequence
```

Comment:
This is not nessesary anymore. The newest version of the detection script handles it via a flag in the decoding step.